lib/grandpa/vote_message.go

```diff
@@ -20,58 +20,49 @@ import (
 	"bytes"
 	"context"
 	"errors"
-	"time"
 
 	"github.com/ChainSafe/gossamer/lib/blocktree"
 	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
 	"github.com/ChainSafe/gossamer/lib/scale"
+
+	"github.com/libp2p/go-libp2p-core/peer"
 )
 
+type networkVoteMessage struct {
+	from peer.ID
+	msg  *VoteMessage
+}
+
 // receiveMessages receives messages from the in channel until the specified condition is met
-func (s *Service) receiveMessages(cond func() bool) {
-	ctx, cancel := context.WithCancel(s.ctx)
-
-	go func() {
-		for {
-			select {
-			case msg := <-s.in:
-				if msg == nil {
-					continue
-				}
-
-				logger.Trace("received vote message", "msg", msg)
-				vm, ok := msg.(*VoteMessage)
-				if !ok {
-					continue
-				}
-
-				v, err := s.validateMessage(vm)
-				if err != nil {
-					logger.Debug("failed to validate vote message", "message", vm, "error", err)
-					continue
-				}
-
-				logger.Debug("validated vote message",
-					"vote", v,
-					"round", vm.Round,
-					"subround", vm.Message.Stage,
-					"prevote count", s.lenVotes(prevote),
-					"precommit count", s.lenVotes(precommit),
-					"votes needed", s.state.threshold(),
-				)
-			case <-ctx.Done():
-				logger.Trace("returning from receiveMessages")
-				return
+func (s *Service) receiveMessages(ctx context.Context /*cond func() bool*/) {
+	for {
+		select {
+		case msg := <-s.in:
```

Comment:
Could add a check to the closed channel? 
```go
case msg, ok := <-s.in:
  if !ok { ... }
```

Reply:
done!